Call updates and multi-use consumables

jbossjaslow opened the pull request:
Overhauled the Call function to better handle mod items and streamline the process for developers. Also added multi-use consumables as a config option

jbossjaslow committed:
Call updates and multi-use consumables

Overhauled the Call function to better handle mod items and streamline the process for developers. Also added multi-use consumables as a config option

direwolf420 commented on Models/Consumable.cs:
```diff
@@ -1,79 +1,80 @@
 
-		private void SetTextureAndDimensions(string texturePath, (float X, float Y) assetDimensions) {
+		#region Private helpers
+		private void SetTextureAndDimensions(string texturePath) {
```
Can be renamed to just `SetTexture`. Same with the other method of the same name

direwolf420 commented on Models/Consumable.cs:
```diff
@@ -1,79 +1,80 @@
-			HasBeenConsumed = hasBeenConsumed;
+		#region Init
+		public Consumable(ModItem modItem, object CategoryOrCustomColor, Func<int> currentCount, Func<int> totalCount) {
+			DisplayText = modItem.DisplayName.ToString();
+			HoverText = modItem.Tooltip.ToString();
```
This is not solving the localization problem outlined in the issue. Invoking .Tostring() or .Value fetches the string for the currently selected language. Changing language then won't change these until next mod reload. To remedy this, don't call these methods, and instead store DisplayText/HoverText as a `LocalizedText`

direwolf420 commented on Models/Consumable.cs:
```diff
@@ -1,79 +1,80 @@
-
-			SetTextureAndDimensions(texturePath: texturePath, assetDimensions: assetDimensions);
+		public Consumable(int vanillaItemId, ConsumableType type, Func<int> currentCount, Func<int> totalCount) {
+			DisplayText = Lang.GetItemName(vanillaItemId).Value;
+			HoverText = Language.GetText($"ItemTooltip.{ItemID.Search.GetName(vanillaItemId)}").Value;
```
Same as above

direwolf420 commented on Models/ConsumableMod.cs:
```diff
@@ -12,10 +14,32 @@ public class ConsumableMod {
+			SetTextureAndDimensionsForExternalMod(mod);
+		}
+
+		private void SetTextureAndDimensionsForExternalMod(Mod mod) {
```
`SetTextureForExternalMod`

direwolf420 commented on Models/ConsumableMod.cs:
```diff
@@ -12,10 +14,32 @@ public class ConsumableMod {
+					ModTabTexture = texture;
+					UsingMissingTexture = false;
+				} else {
+					mod.Logger.Info("icon_small needs to be 30x30 pixels");
```
The logging is a relic of tml copypaste, should be removed. It should be mentioned in the documentation for the calls that icon_small is used for the tab. It will encourage modders to add it themselves

direwolf420 commented on Models/Report.cs:
```diff
@@ -10,62 +10,37 @@ public class Report {
-					entry.Consumables.Add(consumable);
-					return true;
+			foreach (Consumable c in entry.Consumables) {
+				if (c.DisplayText == consumable.DisplayText) {
+					// consumable already exists, exit
```
Probably better to store a unique identifier (such as the item's type (ModItem.Type / id) to then check for equality. This may also cause collisions if two mods add an item with the same name

direwolf420 commented on Models/VanillaConsumable.cs:
```diff
@@ -10,6 +10,11 @@ public VanillaConsumable(ConsumableItem item) {
 		}
 
 		public ConsumableType Type => Item switch {
```
This whole class has alot of hardcode which I don't think is necessary. You can compress the DisplayText to just `Lang.GetItemName(vanillaItemId)`.
Not sure about the HoverText since you add some extra stuff to it. In general, any strings you are using that should be displayed to users are supposed to go into the localization file (like the config does at the moment). You can read more about how to add localization to your mod here: <https://github.com/tModLoader/tModLoader/wiki/Localization>.
ConsumableItem -> ItemId is backwards design imo, it should be sufficient to just have them referenced as the ItemId directly as it's vanilla, and use the consts (`ItemID.LifeCrystal` instead of 29 etc.)
My impression is that all of these could be replaced by using the `Consumable` class which you already have. You can use the API in your own mod aswell!

direwolf420 commented on Munchies.cs:
```diff
@@ -29,112 +29,85 @@ public override void Unload() {
+
+				// Add optional fields for max # of items, option of func<int> for how many have been consumed
+			} catch (Exception e) {
+				(mod ?? this).Logger.Error($"Error adding consumable: {e.Message}, {e.StackTrace}");
```
Two things:
1. The logger should still originate from your own mod always (`this`), since you invoke the logging, and the mod's Name should be used in the log message to indicate/point to it
2. Logging the exception's data is redundant as TML hooks into the exception logging which bypasses catching, so it always prints these out

direwolf420 commented on Munchies.cs:
```diff
@@ -29,112 +29,85 @@ public override void Unload() {
+
+				return Report.AddConsumableToList(mod: vanillaMod, consumable: consumable);
+			} catch (Exception e) {
+				Logger.Error($"Error adding consumable: {e.Message}, {e.StackTrace}");
```
Same as above

direwolf420 reviewed (commented):
As an aside, whenever you do `Request<Texture2D>` to use in UIs (for fetching dimensions primarily), you should use it with `ReLogic.Content.AssetRequestMode.ImmediateLoad`. Otherwise it can happen that the texture is not loaded when accessing dimensions and it returns 0, causing issues in the UI

jbossjaslow committed:
PR feedback

Fixed localized text strings
Removed unneeded files
Moved content from VanillaConsumable to Consumable
Cleaned up exception handling
Added item handler to redraw list if item is used that corresponds to one in the list
Refactored ReportUI to only draw item list on first appearance or on tab change

direwolf420 committed:
fix ImmediateLoad use

jbossjaslow committed:
Merge pull request #3 from direwolf420/feature/call-updates

fix ImmediateLoad use

jbossjaslow committed:
More feedback

Added new multi use handler for calls
Changed default value for showing multi use consumables to true in config
Refactored ConsumableMod to not load texture on init, instead func to get texture when needed

jbossjaslow committed:
Small updates

List item text now resizes itself if an item is used that causes it to get a checkmark or update ratio text

jbossjaslow committed:
Last bit of changes

jbossjaslow committed:
Added star icon for completed tabs

direwolf420 committed:
fix extra tooltip

direwolf420 committed:
localized UI

direwolf420 committed:
fix indent

direwolf420 committed:
fix indent 2

jbossjaslow committed:
Merge pull request #4 from direwolf420/feature/call-updates_loc

Update localization

jbossjaslow committed:
Final tweaks

Greyed out checkmarks, added available and difficulty args, added call info to readme, modified localized text

direwolf420 committed:
auto diff and type tooltips

direwolf420 committed:
better ConsumableMod comparison

direwolf420 committed:
make available optional (default true)

direwolf420 committed:
update readme

jbossjaslow committed:
Merge pull request #5 from direwolf420/feature/call-updates_enum_func

Boilerplate localization removal, optional Available

jbossjaslow committed:
Added vanilla multi use call

jbossjaslow committed:
Split colors from category

jbossjaslow committed:
Trying to determine why vanilla item isn't working

jbossjaslow committed:
Finished readme and added small fix to call difficulty handling

jbossjaslow committed:
Last update to readme

jbossjaslow committed:
Update description_workshop.txt
